### ISSUE TITLE
Optimizing over only some parameters

### ISSUE THREAD

**Issue by zackchase** [tensorflow/tensorflow#834]
Currently the optimizers in TensorFlow appear to act simultaneously upon all tf.Variables in the function graph. However, commonly, it's useful to optimize only over certain Variables. For example, given a pre-trained network, I might only want to optimize wrt a new set of output weights. Am I misreading the examples and optimizer code or is this presently not supported?

**Comment by rafaljozefowicz** [comment]
`optimizer.minimize(loss, var_list=[your variables])` does this if I understand correctly.

Also, take a look at: http://stackoverflow.com/questions/34477889/holding-variables-constant-during-optimizer/34478044#34478044

**Comment by zackchase** [comment]
Thanks!

**Comment by vrv** [comment]
@zackchase: in the future I would use StackOverflow for these types of questions.

**Comment by amohant4** [comment]
optimizer.minimize(loss, var_list=[your variables]) will optimize over the list of variables. 

But is there a way to optimize over a subset of a given variable while treating the rest of it as constants? What I mean is : say I have a 2D variable named W_fc of size 2x2. I want to optimize over W_fc[0][0] and W_fc[1][1] and treat W_fc[0][1] and W_fc[1][0] as constants.

**Comment by wenwei202** [comment]
@amohant4 I am also looking for this. Seems [scatter_update](https://www.tensorflow.org/api_docs/python/tf/scatter_update) may work. My concern is its computation efficiency.

**Comment by thangbui** [comment]
@amohant4 have you got a solution for the case above? 

I want to get gradients w.r.t. a subset of a variable, like tf.gradients([loss], [variable[i]] but the returns None as slicing returns a new tensor. This is closely related to this [stackoverflow question](https://stackoverflow.com/questions/41889531/slice-of-a-variable-returns-gradient-none).

**Comment by jeffpollock9** [comment]
@thangbui I found an answer to your question [here](https://stackoverflow.com/questions/49048622/tensorflow-minimise-with-respect-to-only-some-elements-of-a-variable).

**Comment by MishaLaskin** [comment]
You can you `tf.stop_gradient(x)` if you'd like to treat `x` as a constant when performing gradient descent. I found this API especially useful when implementing Q networks in RL where your target (the Bellman backup)  and loss function (the Q network) share the same weights. `stop_gradient` freezes the Bellman backup weights so the target is treated as a constant.

**Comment by AlexTS1980** [comment]
I wonder what exactly it means 'to optimize with respect to some parameters'. Does this mean only the weights in `[var_list]` are updated? Derivatives only wrt variables in `[var_list]` are calculated and the rest are set to 0? In the code I use [var_list] is in fact a layer (output), not a set of weights.
